packages/@expo/cli/src/utils/codesigning.ts

```diff
@@ -0,0 +1,298 @@
+    await getProjectPrivateKeyAndCertificateFromFilePathsAsync({
+      codeSigningCertificatePath,
+      privateKeyPath,
+    });
+
+  return {
+    privateKey: privateKeyPEM,
+    certificateForPrivateKey: certificatePEM,
+    certificateChainForResponse: [],
+  };
+}
+
+async function readFileWithErrorAsync(path: string, errorMessage: string): Promise<string> {
+  try {
+    return await fs.readFile(path, 'utf8');
+  } catch {
+    throw new Error(errorMessage);
+  }
+}
+
+async function getProjectPrivateKeyAndCertificateFromFilePathsAsync({
+  codeSigningCertificatePath,
+  privateKeyPath,
+}: {
+  codeSigningCertificatePath: string;
+  privateKeyPath: string;
+}): Promise<{ privateKeyPEM: string; certificatePEM: string }> {
+  const [codeSigningCertificatePEM, privateKeyPEM] = await Promise.all([
+    readFileWithErrorAsync(
+      codeSigningCertificatePath,
+      `Code signing certificate cannot be read from path: ${codeSigningCertificatePath}`
+    ),
+    readFileWithErrorAsync(
+      privateKeyPath,
+      `Code signing private key cannot be read from path: ${privateKeyPath}`
+    ),
+  ]);
+
+  const privateKey = convertPrivateKeyPEMToPrivateKey(privateKeyPEM);
+  const certificate = convertCertificatePEMToCertificate(codeSigningCertificatePEM);
+  validateSelfSignedCertificate(certificate, {
+    publicKey: certificate.publicKey as PKI.rsa.PublicKey,
+    privateKey,
+  });
+
+  return { privateKeyPEM, certificatePEM: codeSigningCertificatePEM };
+}
+
+function validateStoredDevelopmentExpoRootCertificateCodeSigningInfo(
+  codeSigningInfo: StoredDevelopmentExpoRootCodeSigningInfo,
+  easProjectId: string
+): CodeSigningInfo | null {
+  if (codeSigningInfo.easProjectId !== easProjectId) {
+    return null;
+  }
+
+  const { privateKey: privateKeyPEM, certificateChain: certificatePEMs } = codeSigningInfo;
+  if (!privateKeyPEM || !certificatePEMs) {
+    return null;
+  }
+
+  const certificateChain = certificatePEMs.map((certificatePEM) =>
+    convertCertificatePEMToCertificate(certificatePEM)
+  );
+
+  // TODO(wschurman): maybe move to @expo/code-signing-certificates
+  const leafCertificate = certificateChain[0];
+  const now = new Date();
+  if (leafCertificate.validity.notBefore > now || leafCertificate.validity.notAfter < now) {
+    return null;
+  }
+
+  // TODO(wschurman): maybe do more validation
+
+  return {
+    certificateChainForResponse: certificatePEMs,
+    certificateForPrivateKey: certificatePEMs[0],
+    privateKey: privateKeyPEM,
+  };
+}
+
+async function fetchAndCacheNewDevelopmentCodeSigningInfoAsync(
+  projectRoot: string,
+  easProjectId: string
+): Promise<CodeSigningInfo> {
+  const keyPair = generateKeyPair();
+  const keyPairPEM = convertKeyPairToPEM(keyPair);
+  const csr = generateCSR(keyPair, `Development Certificate for ${easProjectId}`);
+  const csrPEM = convertCSRToCSRPEM(csr);
+  const [developmentCertificate, expoGoIntermeidateCertificate] = await Promise.all([
```

Comment:
Putting "signing" in the name makes it clearer that it's the leaf certificate doing the actual signing:
```suggestion
  const [developmentSigningCertificate, expoGoIntermediateCertificate] = await Promise.all([
```